### DISCUSSION TITLE
Image Masking in v 2023

### DISCUSSION THREAD

**Discussion by jgmenzer** [alicevision/Meshroom#1993] [Q&A] [answered]
Hi All,

I am curious if anyone can provide some documentation/help using the ImageMasking node (or best practice outside of this node)? I combed through some old posts, but the seemed they were before the ImageMasking node was implemented.

My set up: using a turntable with reference sheet (sam porter's) inside a photo box. I can change out the colors, but would prefer black or white as background.

Questions: What is the input/output for this node (e.g. where in the pipeline should it go? If using black/white background should I be using HSV or greyscale? If using HSV, I see parameters for green, white, or black, but not sure which Hue value to use (or maybe it doesn't matter here)?

I appreciate any help, and apologize if this is redundant from somewhere else, Jeremy

**Comment by natowi** [comment, marked as answer]
You can get the setting by hovering over the parameter (but yes, the hue info is missing):
HSV
Green : default values
White : Hue=AnyValue, Tolerance=1, minSaturation=0, maxSaturation=0.1, minValue=0.8, maxValue=1
Black : Hue=AnyValue, Tolerance=1, minSaturation=0, maxSaturation=0.1, minValue=0, maxValue=0.2

The Input from the ImageMasking connects to CameraInit.
The output of the ImageMasking node can be connected to FeatureExtraction (use mask for in camera estimation) Masking Input and the PrepareDenseScene Masking Input (applies the undistort to the mask and combines it with the RGB to a RGBA image).

![grafik](https://user-images.githubusercontent.com/8138537/234055608-94abab58-e947-4f49-9e46-3935af98ebb9.png)


There are also  some new features for marker detection that will be useful (I need to write up some text for it ;) )

**Reply by jgmenzer** [reply]
Thanks for the response and explanation natowi!

Yes, marker detection was my next step. I am currently modifying Sam Porter's reference sheet (in photo) to add the CCTAG3 markers. From what I can tell so far there isn't a way to use other marker tags (the black and white circle's A,B,C,D)? Also, I am guessing I will change out the background to white since my turntable is white, not sure if this will make a big difference or not?

Thanks so much!

![DSC_0006](https://user-images.githubusercontent.com/33295507/234097575-0888dd64-e975-4d3e-bcfd-17dcf6b1c393.JPG)

**Reply by natowi** [reply]
The A B C D are just generic markers for manual use.

Take a look at the tutorials here: https://github.com/alicevision/Meshroom/wiki/Tutorials
-> Scanning in the Void - A short video on good turntable setups https://www.youtube.com/watch?v=Il6LVXqSlRg

https://github.com/alicevision/Meshroom/wiki/CCTAG-scaling

Here is an example for a white reference sheet: https://www.archeo3d.net/?page_id=3810 (sadly no pdf)
tag16h5 are also supported with the latest release btw.

**Reply by jgmenzer** [reply]
Thanks for all the great information, much appreciated!

**Reply by jgmenzer** [reply]
@natowi is there a way to use markers (cctag3) while image masking white/black, or do I need to use a different color e.g. green? When masking out white, no markers are detected, which makes sense given the cctag3 markers are predominantly white. I haven't tested, but guessing this would also be the case for black. I did test feature extraction with and without the image masking, with no white mask all cctags are detected.

Scanning in the void looks awesome, but I don't have the right equipment for that unfortunately.

Thanks, Jeremy

**Reply by natowi** [reply]
@jgmenzer you can do the cctag detection with a parallel  node similar to the workflow described in the section "Improve CCTAG detection" here https://github.com/alicevision/Meshroom/wiki/CCTAG-scaling

**Reply by natowi** [reply]
@jgmenzer Here you go :)

![grafik](https://user-images.githubusercontent.com/8138537/234973086-538c48dd-9ef9-4c3d-a257-fc577c90db5e.png)

Here is what I did:

- add the ImageMasking node to the default node graph
- create a parallel branch for CCTAG detection only
- use both branches as input in the StructureFromMotion node

To check the FeatureExtraction results double click on the node. Disable the preview before double clicking on the node from the other branch, so the viewer updates correctly

*in the node configuration Note Tab you can rename your nodes and add comments or a custom colour

I was only able to do some rough testing, so I´d be happy to hear if this worked for you

Here is the Meshroom.mg file content (copy-paste in text file and save it as .mg. Then in Meshroom go to File->Advanced->Import Template):
https://gist.github.com/natowi/52eb673a75f9fbf6181b86bd879ada5a

**Reply by jgmenzer** [reply]
@natowi Thanks so much, sorry for my delay in getting back to this. So that all runs just fine, but, I then added a SfM transform node (to orient based on the cctags.  SfM > SfM Transform (Transformation method - from_markers, Landmark describer types - (tried dpsift and cctag3, and just cctag3), added my marker information. I get [error] Alignment from markers: Invalid number of image describer types: 0.

I am probably missing something simple here. 

Jeremy

**Reply by natowi** [reply]
Hi Jeremy, could you share a screenshot of your graph and the sfm transform settings?

**Reply by jgmenzer** [reply]
Hey natowi - I've used your template (no settings changed). Then added the SfMTransform. 

![image](https://github.com/alicevision/Meshroom/assets/33295507/d2bd50d0-b8b1-42fb-afb1-3678ec75e171)

For SfMTransform: Transform Method - from_markers, Landmark Describer types -cctag3 (I also tried cctag3 and dpsift). My Marker positions, all others default.

![image](https://github.com/alicevision/Meshroom/assets/33295507/cb5acd11-50c0-4afa-9c11-f95d7f66b6e0)
![image](https://github.com/alicevision/Meshroom/assets/33295507/a6ba1c78-1111-4bbf-9005-3e45b46ee122)
![image](https://github.com/alicevision/Meshroom/assets/33295507/3e5a312b-2cd4-4214-b39b-c48eaae18787)

Not sure if this will help:
![image](https://github.com/alicevision/Meshroom/assets/33295507/5ee53dad-0553-4487-89ee-2c9c1a07c8ce)

Thanks, Jeremy

**Reply by natowi** [reply]
Did you try with dspsift and cctag checked at the same time?

//Try duplicating the StructureFromMotion node and use sfmTransform after the FeatureMatching CCTAG node, then add SfM Alignment node and use the CCTAG input as reference and the StructureFromMotion as input, use the output to continue.
Something in this direction:
![grafik](https://github.com/alicevision/Meshroom/assets/8138537/df511f63-fb64-47d5-8b6f-849cefa837a7)

**Reply by jgmenzer** [reply]
Yes I tried dpsift and cctag checked at the same time, same error. 

Trying what you suggest directly above is working (I actually tried this earlier, but had poses as reference not sfmData file.....). For alignment method, I used "from_cameras_filepath" all other methods fail to find similarity between the two scenes. Does this make sense to you/sound correct?

Thanks so much!!! This should work for me now. 

Also FYI - 
Not sure if I should open a different issue. I am running an augmentation after this process to add the bottom of a lithic artifact to the model (covered by clay to hold it up). So with the white masking background, the augmentation works well and I get a model of the complete object. I also tried with black background (it fails in the augmented SfM) and with a green background, so I wouldn't have to worry about masking out the cctags, that also fails in the augmented sfm. I thought that was odd as the other colors are masked out and I get similar numbers of features and landmarks detect. Thoughts here?

**Reply by natowi** [reply]
Good to hear that it works for you.

You can open a new issue for your other question.

//
Note: sfm transform may have some bugs: https://github.com/alicevision/Meshroom/issues/1994
